File: python3/Two Pointers/largest_container.py
from typing import List


def largest_container(heights: List[int]) -> int:
    max_water = 0
    left, right = 0, len(heights) - 1
    while (left < right):
        # Calculate the water contained between the current pair of 
        # lines.
        water = min(heights[left], heights[right]) * (right - left)
        max_water = max(max_water, water)
        # Move the pointers inward, always moving the pointer at the 
        # shorter line. If both lines have the same height, move both 
        # pointers inward.
        if (heights[left] < heights[right]):
            left += 1
        elif (heights[left] > heights[right]):
            right -= 1
        else:
            left += 1
            right -= 1
    return max_water

File: kotlin/Two Pointers/LargestContainer.kt
class LargestContainer {

    fun largestContainer(heights: List<Int>): Int {
        var maxWater = 0
        var left = 0
        var right = heights.size - 1
        while (left < right) {
            // Calculate the water contained between the current pair of
            // lines.
            val water = min(heights[left], heights[right]) * (right - left)
            maxWater = max(maxWater, water)
            // Move the pointers inward, always moving the pointer at the
            // shorter line. If both lines have the same height, move both
            // pointers inward.
            if (heights[left] < heights[right]) {
                left++
            } else if (heights[left] > heights[right]) {
                right--
            } else {
                left++
                right--
            }
        }
        return maxWater
    }
}